R/pls-mvr-tidiers.R

```diff
@@ -0,0 +1,203 @@
+#'   "response",
+#'   ".fitted",
+#'   .t.squared="Hotelling's T^2 value."
+#' )
+#'
+#' @inherit tidy.mvr examples
+#'
+#' @family mvr tidiers
+#'
+#' @export
+#'
+#' @seealso [augment()], [pls::predict.mvr()]
+#'
+augment.mvr <- function(x,
+                        data = model.frame(x),
+                        newdata = NULL,
+                        ncomp = x$ncomp, ...) {
+  if (length(ncomp) > 1) {
+    line <- paste(
+      "ncomp is a multi-component vector:",
+      "ncomp = {ncomp[1]} will be used."
+    )
+    message(glue(line))
+  }
+  temp <- x
+  class(temp) <- c("mvr_temp", class(temp))
+  df <- augment_newdata(temp,
+    data,
+    newdata,
+    FALSE,
+    ncomp = ncomp[1],
+    type = "response", ...
+  )
+  df$.scores <- predict(x, newdata = newdata, type = "scores", ...)
+
+  t2 <- as.data.frame(df$.scores^2)
+  t2 <- purrr::map2(t2, x$Xvar, ~ .x / .y)
+  t2 <- purrr::accumulate(t2, `+`)
+  t2 <- as.data.frame(t2)
+  t2 <- tibble::tidy_names(t2)
+  t2 <- as.matrix(t2)
+  df$.t.squared <- t2
+
+  df
+}
+
+# Make predict(), an S3 method, preferable in augment_newdata() using a temporary class
+predict.mvr_temp <- function(object, ncomp, type, ...) {
+  pls:::predict.mvr(object, ncomp = ncomp, type = type, ...)[, , 1]
+}
+
+#' @templateVar class mvr
+#' @template title_desc_glance
+#'
+#' @param x An `mvr` object such as those created by [pls::mvr()], [pls::plsr()], [pls::cppls()]
+#'  and [pls::pcr()].
+#' @param ncomp The number of components to include in the model. Ignored if comps is specified.
+#' @param test A [data.frame()] or [tibble::tibble()] to estimate the predictive performance of 
+#' `x`. It contains all the original predictors used to create `x`.
+#' @param estimate Which estimators to use to estimate mesures of fit, containing the mean
+#'  squared error of prediction (MSEP) root mean squared error of prediction (RMSEP) and R^2.
+#'   "adjCV" is only available for (R)MSEP.
+#' @template param_unused_dots
+#'
+#' @evalRd return_glance(
+#'   "estimator",
+#'   "r.squared",
+#'   msep = "Mean squared error of prediction",
+#'   rmsep = "Root mean squared error of prediction"
+#' )
+#'
+#' @inherit tidy.mvr examples
+#'
+#' @family mvr tidiers
+#'
+#' @export
+#'
+#' @seealso [glance()], [pls::R2()]
+#'
+glance.mvr <- function(x, 
+                       ncomp = x$ncomp, 
+                       test = NULL,
+                       estimate = c("train", "test", "CV", "adjCV"), ...) {
+  if (length(ncomp) > 1) {
+    line <- paste(
+      "ncomp is a multi-component vector:",
+      "ncomp = {ncomp[1]} will be used."
+    )
+    message(glue(line))
+  }
+  
+  estimate <- match.arg(estimate)
+  if (!is.null(test) && estimate != "test") {
+    message("test is not NULL: estimate = \"test\" will be used.")
+    estimate <- "test"
+  }
+  
+  ret <- tibble::tibble(estimator = estimate, r.squared = NA)
+  if (estimate != "adjCV") ret$r.squared <- grance.mvr.estimate(x, pls::R2, ncomp[1], test, estimate)
```

Comment:
This `do.call()` helper strategy will be hard to maintain down the line. Can we replace this with explicit calls to `pls::MSEP`, `pls::RMSEP`, etc?